### PR TITLE
feat: make tweets be automatically deleted

### LINKED ISSUE #13: Delete Unavailable Tasks From Twitter
Delete tweets for tasks that are assigned or closed!

>> Just newly posted ones should be tweeted.

> Should we delete a tweet if an issue was closed?

_Originally posted by @rndquu in https://github.com/ubiquity/devpool-directory/pull/1023#issuecomment-1909430199_

### PR DESCRIPTION
**Pull request by BeanieMen**
Resolves [#13](https://github.com/ubiquity/devpool-directory-bounties/issues/13)

### CODE REVIEW

**Comment by BeanieMen** [comment]
How should i qa this

**Comment by BeanieMen** [comment]
Idk how to qa this tbh

**Comment by gitcoindev** [comment]
@BeanieMen perhaps this can help https://github.com/twitterdev/twitter-api-typescript-sdk/blob/main/test/v2.client.test.ts ?

**Comment by EtherealGlow** [comment]
Ooo thanks mr gitcoindev

**Comment by BeanieMen** [comment]
Oo thanks mr gitcoindev

**Comment by BeanieMen** [comment]
but this would still require the twitter tokens

**Comment by BeanieMen** [comment]
@rndquu how do you think this should be tested

**Comment by rndquu** [comment]
> @rndquu how do you think this should be tested

It should be tested on your local twitter account, you may apply for the twitter's dev account and get an API token, I got mine in a minute

**Comment by BeanieMen** [comment]
I didn't even try to make one because Pavlovcik said it was tedious 💀

**Comment by rndquu** [comment]
> I didn't even try to make one because Pavlovcik said it was tedious 💀

Perhaps for twitter api v1 it was tedious, but since v2 getting an API key seems to be simpler

**Comment by BeanieMen** [comment]
```
  data: {
    client_id: '28388546',
    detail: 'When authenticating requests to the Twitter API v2 endpoints, you must use keys and tokens from a Twitter developer App that is attached to a Project. You can create a project via the developer portal.',
    registration_url: 'https://developer.twitter.com/en/docs/projects/overview',
    title: 'Client Forbidden',
    required_enrollment: 'Appropriate Level of API Access',
    reason: 'client-not-enrolled',
    type: 'https://api.twitter.com/2/problems/client-forbidden'
  }
```
Seems like my twitter app is forbidden from making requests of type `userTimeline`.
Can yall test with your twitter app?

**Comment by BeanieMen** [comment]
```
let { data } = await twitterClient.v2.userTimeline(userId, { max_results: 100 });
```
heres the offending method

**Comment by rndquu** [comment]
> ```
>   data: {
>     client_id: '28388546',
>     detail: 'When authenticating requests to the Twitter API v2 endpoints, you must use keys and tokens from a Twitter developer App that is attached to a Project. You can create a project via the developer portal.',
>     registration_url: 'https://developer.twitter.com/en/docs/projects/overview',
>     title: 'Client Forbidden',
>     required_enrollment: 'Appropriate Level of API Access',
>     reason: 'client-not-enrolled',
>     type: 'https://api.twitter.com/2/problems/client-forbidden'
>   }
> ```
> 
> Seems like my twitter app is forbidden from making requests of type `userTimeline`. Can yall test with your twitter app?

1. This is your job to make sure the PR is working fine
2. I think your should follow what the `detail` field states: create a developer account, create a project, create an app in that project and use those API keys

**Comment by BeanieMen** [comment]
I already created the app in the default project, it only states that I am forbidden from making requests for timeline so I thought it was a access level issue

**Comment by rndquu** [comment]
> I already created the all in the default project, it only states that I am forbidden from making requests for timeline so I thought it was a access level issue

If you get an access level error then we'll get the same one. You should find out what's wrong and provide a solution.

**Comment by BeanieMen** [comment]
@rndquu I looked into it, the only endpoints which are available for the free tier posting tweets and deleting tweets
```
POST /2/tweets
DELETE /2/tweets/:id
```
I was using timelines to get all tweets for a user which isnt available.
I dont think we can search and delete tweets using the twitter api.
If you have any thoughts on any alternate idea. Pls tell

**Comment by BeanieMen** [comment]
@rndquu

**Comment by rndquu** [comment]
> @rndquu I looked into it, the only endpoints which are available for the free tier posting tweets and deleting tweets
> 
> ```
> POST /2/tweets
> DELETE /2/tweets/:id
> ```
> 
> I was using timelines to get all tweets for a user which isnt available. I dont think we can search and delete tweets using the twitter api. If you have any thoughts on any alternate idea. Pls tell

Fetching all twitter posts will eventually exceed API rate limits.

We could:
1. On every posted tweet save it's id to a key/value mapping json file, like:
```
{
  "github_issue_node_id": "twitter_post_id"
}
```
2. Save this json file to github artifacts
3. When some issue is closed then search through that json file for tweet id and call https://developer.twitter.com/en/docs/twitter-api/tweets/manage-tweets/api-reference/delete-tweets-id

**Comment by BeanieMen** [comment]
Wasn't able to work yesterday because of reasons, sorry bout that

**Comment by BeanieMen** [comment]
will qa in a bit

**Comment by BeanieMen** [comment]
tweet posted successfully https://github.com/BeanieMen/devpool-directory/actions/runs/7940842664/job/21682556132

**Comment by BeanieMen** [comment]
tweet deleted successfully also https://github.com/BeanieMen/devpool-directory/actions/runs/7940849501/job/21682570537

**Comment by BeanieMen** [comment]
@rndquu can you review this?

File: helpers/twitter.ts [hunk 1]
```diff
@@ -33,3 +34,16 @@ async function postTweet(status: string) {
     console.error("Error posting tweet", error);
   }
 }
+
+async function deleteTweet(id: string) {
+  try {
+    const { data } = await twitterClient.v2.deleteTweet(id);
+    if (data.deleted) {
```
Comment (0x4007) [inline helpers/twitter.ts:41]: Pretty sure there's a bug here

https://github.com/BeanieMen/devpool-directory/actions/runs/7940849501/job/21682570537#step:7:12

Comment (BeanieMen) [reply helpers/twitter.ts:41]: that was a redundant log removed in https://github.com/ubiquity/devpool-directory/pull/1046/commits/e87aecb79c0451efcc7e08b317da3712568ec396

Comment (BeanieMen) [reply helpers/twitter.ts:41]: its showing undefined because i logged it after deleting key-value pair from twitterMap (ik i am dumb)

**Comment by rndquu** [comment]
> @rndquu can you review this?

Hey, I'll check it today

**Review by rndquu — CHANGES_REQUESTED** [review]
Getting this CI error on syncing https://github.com/rndquu/devpool-directory/actions/runs/7976309853/job/21776586613

There is no `twitterMap` artifact initially hence the workflow fails

**Comment by BeanieMen** [comment]
@rndquu can I get a review here?

File: index.ts [hunk 2]
```diff
@@ -101,7 +117,7 @@ async function main() {
         } else {
           // issue does not exist in devpool
           // if issue is "closed" then skip it, no need to copy/paste already "closed" issues
-          if (projectIssue.state === "closed") continue;
+          if (projectIssue.state === "closed" || projectIssue.assignee?.login) continue;
```
Comment (rndquu) [inline index.ts]: @BeanieMen Current issue is about deleting tweets. Why did you make this change?

**Comment by BeanieMen** [comment]
Why am I so fucking dumb 😭

**Comment by BeanieMen** [comment]
I mixed up em branches ffs

**Comment by BeanieMen** [comment]
One day I am gonna get as good as rndquu and 0xcodercrane

File: index.ts (hunk 2 above)
Comment (BeanieMen) [reply index.ts]: I just noticed even Alex is laughing at me 😭

Comment (BeanieMen) [reply index.ts]: Also I fixed the dumbass mistake

**Review by rndquu — APPROVED** [review]
Works fine

**Comment by 0x4007** [comment]
@rndquu should we/can you merge this in if everything checks out?

**Pull request merged**